### PR TITLE
rpcdaemon: use boost::beast for HTTP

### PR DESCRIPTION
**Pull request by lupin012**

### CODE REVIEW

File: silkworm/rpc/test/api_test_database.hpp [hunk 1]
```diff
@@ -53,10 +53,19 @@ InMemoryState populate_genesis(db::RWTxn& txn, const std::filesystem::path& test
...
     Task<void> open_stream() override { co_return; }
-    Task<void> write_rsp(Response& /* response */) override { co_return; }
+    Task<void> write_rsp(std::string& response) override {
+        response_ = response;
+        co_return;
+    }
     Task<std::size_t> write(std::string_view /* content */) override { co_return 0; }
     Task<void> close() override { co_return; }
+
+    std::string& get_response() { return response_; }
```
Comment (canepat) [inline silkworm/rpc/test/api_test_database.hpp:65]: `std::string&` -> `const std::string&`

File: silkworm/rpc/test/api_test_database.hpp [hunk 2]
```diff
@@ -53,10 +53,19 @@ InMemoryState populate_genesis(db::RWTxn& txn, const std::filesystem::path& test
 void populate_blocks(db::RWTxn& txn, const std::filesystem::path& tests_dir, InMemoryState& state_buffer);
 
 class ChannelForTest : public Channel {
+  public:
     Task<void> open_stream() override { co_return; }
-    Task<void> write_rsp(Response& /* response */) override { co_return; }
+    Task<void> write_rsp(std::string& response) override {
```
Comment (canepat) [inline silkworm/rpc/test/api_test_database.hpp:58]: `std::string&` -> `const std::string&`

File: silkworm/rpc/http/request_handler.cpp [hunk 3]
```diff
@@ -33,17 +33,16 @@ constexpr std::size_t kStreamBufferSize{4096};
 
 Task<void> RequestHandler::handle(const std::string& content) {
     auto start = clock_time::now();
-    Channel::Response msg_response;
+    std::string response{};
```
Comment (canepat) [inline silkworm/rpc/http/request_handler.cpp:36]: No need to initialise empty string

File: silkworm/rpc/http/connection.cpp [hunk 4]
```diff
@@ -45,13 +46,8 @@ Connection::Connection(boost::asio::io_context& io_context,
                        std::optional<std::string> jwt_secret)
     : socket_{io_context},
       request_handler_{this, api, handler_table},
-      buffer_{},
       allowed_origins_{allowed_origins},
       jwt_secret_{std ::move(jwt_secret)} {
-    request_.content.reserve(kRequestContentInitialCapacity);
-    request_.headers.reserve(kRequestHeadersInitialCapacity);
-    request_.method.reserve(kRequestMethodInitialCapacity);
-    request_.uri.reserve(kRequestUriInitialCapacity);
```
Comment (canepat) [inline silkworm/rpc/http/connection.cpp:54]: The definitions of all these constants (here and in `constants.hpp`) should be removed as they are not used anymore.

File: silkworm/rpc/http/connection.cpp [hunk 5]
```diff
@@ -62,213 +58,172 @@ Connection::~Connection() {
...
+        bytes_transferred = co_await boost::beast::http::async_read(socket_, data_, parser, boost::asio::use_awaitable);
+    } catch (const boost::system::system_error& se) {
+        if (se.code() == boost::beast::http::error::end_of_stream || se.code() == boost::asio::error::broken_pipe) {
+            throw;
+        } else {
+            co_return;
+        }
+    }
+
+    SILK_DEBUG << "Connection::do_read bytes_read: " << bytes_transferred;
```
Comment (canepat) [inline silkworm/rpc/http/connection.cpp:94]: `SILK_DEBUG` -> `SILK_TRACE` in this log line

File: silkworm/rpc/http/connection.cpp [hunk 6]
```diff
@@ -62,213 +58,172 @@ Connection::~Connection() {
...
+    SILK_TRACE << "Connection::do_read: " << parser.get() << "\n";
```
Comment (canepat) [inline silkworm/rpc/http/connection.cpp:95]: I would remove this log line or collapse it with the previous one

File: silkworm/rpc/http/request_handler.cpp [hunk 7]
```diff
@@ -57,19 +56,18 @@ Task<void> RequestHandler::handle(const std::string& content) {
...
         batch_reply_content << "]\n";
 
-        msg_response.status = Channel::ResponseStatus::ok;
-        msg_response.content = batch_reply_content.str();
+        response = batch_reply_content.str();
     }
 
     if (send_reply) {
-        co_await channel_->write_rsp(msg_response);
+        co_await channel_->write_rsp(response);
```
Comment (canepat) [inline silkworm/rpc/http/request_handler.cpp:70]: I suppose you can move `response` here if `Channel::write_rsp` accepts a r-value parameter

File: silkworm/rpc/http/connection.hpp [hunk 8]
```diff
@@ -60,55 +60,40 @@ class Connection : public Channel {
...
     Task<void> close() override { co_return; }
 
   private:
     using AuthorizationError = std::string;
     using AuthorizationResult = tl::expected<void, AuthorizationError>;
-    AuthorizationResult is_request_authorized(const http::Request& request);
+    AuthorizationResult is_request_authorized(boost::beast::http::request<boost::beast::http::string_body>& req);
 
-    Task<void> handle_request(Request& request);
+    Task<void> handle_request(boost::beast::http::request<boost::beast::http::string_body>& req);
```
Comment (canepat) [inline silkworm/rpc/http/connection.hpp:73]: The `req` parameter should be either `const T&` or `T&&` here

File: silkworm/rpc/http/connection.hpp [hunk 9]
```diff
@@ -60,55 +60,40 @@ class Connection : public Channel {
...
+    Task<void> write_rsp(std::string& response) override;
     Task<void> open_stream() override;
     Task<std::size_t> write(std::string_view content) override;
     Task<void> close() override { co_return; }
 
   private:
     using AuthorizationError = std::string;
     using AuthorizationResult = tl::expected<void, AuthorizationError>;
-    AuthorizationResult is_request_authorized(const http::Request& request);
+    AuthorizationResult is_request_authorized(boost::beast::http::request<boost::beast::http::string_body>& req);
```
Comment (canepat) [inline silkworm/rpc/http/connection.hpp:71]: The `req` parameter should be either `const T&` or `T&&` here

**Review by canepat — CHANGES_REQUESTED** [review]

File: silkworm/rpc/http/request_handler.cpp (hunk 3 above)
Comment (lupin012) [reply silkworm/rpc/http/request_handler.cpp:36]: ok, I have understood that tidy_clang request it

File: silkworm/rpc/http/connection.cpp (hunk 4 above)
Comment (lupin012) [reply silkworm/rpc/http/connection.cpp:54]: ok

File: silkworm/rpc/test/api_test_database.hpp (hunk 2 above)
Comment (lupin012) [reply silkworm/rpc/test/api_test_database.hpp:58]: ok

File: silkworm/rpc/test/api_test_database.hpp (hunk 1 above)
Comment (lupin012) [reply silkworm/rpc/test/api_test_database.hpp:65]: ok

File: silkworm/rpc/http/channel.hpp [hunk 10]
```diff
@@ -22,39 +22,14 @@ namespace silkworm::rpc {
...
-
     Channel() = default;
     ~Channel() override = default;
 
     Channel(const Channel&) = delete;
     Channel& operator=(const Channel&) = delete;
 
     virtual Task<void> open_stream() = 0;
-    virtual Task<void> write_rsp(Response& response) = 0;
+    virtual Task<void> write_rsp(std::string& content) = 0;
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/rpc/http/channel.hpp:32]: I have changed in const std::string& as indicated two comments below

File: silkworm/rpc/http/connection.cpp (hunk 5 above)
Comment (lupin012) [reply silkworm/rpc/http/connection.cpp:94]: ok

File: silkworm/rpc/http/connection.cpp (hunk 6 above)
Comment (lupin012) [reply silkworm/rpc/http/connection.cpp:95]: ok

File: silkworm/rpc/http/request_handler.cpp (hunk 7 above)
Comment (lupin012) [reply silkworm/rpc/http/request_handler.cpp:70]: ok

File: silkworm/rpc/http/connection.hpp (hunk 8 above)
Comment (lupin012) [reply silkworm/rpc/http/connection.hpp:73]: ok, i have declared constT&

File: silkworm/rpc/http/connection.hpp (hunk 9 above)
Comment (lupin012) [reply silkworm/rpc/http/connection.hpp:71]: ok

**Review by canepat — APPROVED** [review]

**Pull request merged**